src/query/storage/m3/dynamic_cluster.go

```diff
@@ -88,10 +96,397 @@ func NewDynamicClusterOptions() DynamicClusterOptions {
 	}
 }
 
-type dynamicCluster struct{}
+type dynamicCluster struct {
+	clusterCfgs []DynamicClusterNamespaceConfiguration
+	logger      *zap.Logger
+	iOpts       instrument.Options
+
+	sync.RWMutex
+
+	namespaces              ClusterNamespaces
+	unaggregatedNamespace   ClusterNamespace
+	aggregatedNamespaces    map[RetentionResolution]ClusterNamespace
+	namespacesByEtcdCluster map[int]clusterNamespaceLookup
+
+	nsWatches []namespace.NamespaceWatch
+	closed    bool
+}
 
 // NewDynamicClusters creates an implementation of the Clusters interface
 // supports dynamic updating of cluster namespaces.
 func NewDynamicClusters(_ DynamicClusterOptions) (Clusters, error) {
 	return nil, errors.New("dynamic cluster configuration not yet supported")
 }
+
+// TODO(nate): Replace constructor above with this method
+// once namespace staging is complete.
+//
+// newDynamicClusters creates an implementation of the Clusters interface
+// supports dynamic updating of cluster namespaces.
+func newDynamicClusters(opts DynamicClusterOptions) (Clusters, error) {
```

Comment:
nit: can probably combine them and have it throw an error early, just so only one line to remove after its staged?

Reply:
Actually left the two constructors in place, so that I can have a different entry point for tests where I do want to flex this functionality.